backend/src/main/java/org/cryptomator/hub/api/UsersResource.java

```diff
@@ -290,4 +324,231 @@ public static TrustedUserDto fromEntity(EffectiveWot entity) {
+			return Response.status(Response.Status.CONFLICT)
+					.entity(e.getMessage())
+					.type(MediaType.TEXT_PLAIN)
+					.build();
+		}
+	}
+
+	@GET
+	@Path("/{id}")
+	@RolesAllowed("admin")
+	@Produces(MediaType.APPLICATION_JSON)
+	@NoCache
+	@Transactional
+	@Operation(summary = "get a specific user")
+	@APIResponse(responseCode = "200", description = "user found")
+	@APIResponse(responseCode = "404", description = "user not found")
+	public UserDtoWithTimestamp getUser(@PathParam("id") String userId) {
+		User user = userRepo.findById(userId);
+		if (user == null) {
+			throw new NotFoundException("User not found: " + userId);
+		}
+
+		Long createdTimestamp = null;
+		String firstName = null;
+		String lastName = null;
+		try {
+			var keycloakUser = keycloakAdminService.getUser(userId);
+			createdTimestamp = keycloakUser.getCreatedTimestamp();
+			firstName = keycloakUser.getFirstName();
+			lastName = keycloakUser.getLastName();
+		} catch (Exception e) {
+			LOG.fine("Could not fetch Keycloak user data for " + userId);
+		}
+
+		// Fetch groups for the user
+		List<GroupDto> groups = userRepo.getGroupsForUser(userId)
+				.map(GroupDto::fromEntity)
+				.toList();
+
+		// Fetch vaults with roles for the user
+		List<VaultResource.VaultDtoWithRole> vaults = userRepo.getVaultAccessForUser(userId)
+				.map(eva -> {
+					Vault vault = vaultRepo.findById(eva.getId().getVaultId());
+					return VaultResource.VaultDtoWithRole.from(vault, eva.getId().getRole());
+				})
+				.toList();
+
+		// Fetch devices (modern devices)
+		Set<DeviceResource.DeviceDto> devices = user.devices.stream()
+				.map(DeviceResource.DeviceDto::fromEntity)
+				.collect(Collectors.toSet());
+
+		// Fetch legacy devices
+		Set<DeviceResource.DeviceDto> legacyDevices = user.legacyDevices.stream()
+				.map(DeviceResource.DeviceDto::fromEntity)
+				.collect(Collectors.toSet());
+
+		// Fetch roles
+		Set<String> roles = keycloakAdminService.getUserRoles(userId);
+
+		return UserDtoWithTimestamp.from(
+				UserDto.justPublicInfo(user),
+				createdTimestamp,
+				firstName,
+				lastName,
+				groups,
+				vaults,
+				devices,
+				legacyDevices,
+				roles
+		);
+	}
+
+	@PUT
+	@Path("/{id}")
+	@RolesAllowed("admin")
+	@Consumes(MediaType.APPLICATION_JSON)
+	@Produces(MediaType.APPLICATION_JSON)
+	@Transactional
+	@Operation(summary = "update a user in Keycloak")
+	@APIResponse(responseCode = "200", description = "user updated")
+	@APIResponse(responseCode = "403", description = "user has federated identity and cannot be modified")
+	@APIResponse(responseCode = "404", description = "user not found")
+	public UserDto updateUser(@PathParam("id") String userId, @Valid @NotNull UpdateUserDto dto) {
+		try {
+			keycloakAdminService.updateUser(
+					userId,
+					dto.firstName(),
+					dto.lastName(),
+					dto.password(),
+					dto.pictureUrl()
+			);
+
+			if (dto.roles() != null) {
+				keycloakAdminService.updateUserRoles(userId, dto.roles());
+			}
```

Comment:
Shouldn't we merge `updateUser()` and `updateUserRoles()` to save a few requests against Keycloak?